### ISSUE TITLE
[Bug] Spark2.5 attention output gate hardcodes sigmoid, ignores config gate_attn_act_mode

### ISSUE THREAD

**Issue by Bevisy** [sgl-project/sglang#37639]
### Checklist

- [x] I searched open and closed issues/PRs and found no existing report for this mismatch.
- [x] The mismatch is present on the latest `main` commit.
- [x] A minimal CPU-only reproduction is included below.
- [x] This is a bug report rather than a general question.
- [x] This report is written in English.

### Describe the bug

The Spark2.5 attention output gate applies a per-head scalar gate computed by `g_proj`, then activates it with a hardcoded `torch.sigmoid`:

[`python/sglang/srt/models/spark2_5.py`](https://github.com/sgl-project/sglang/blob/f8cbf000f4a5bfd86d3fb7c1e2d6c8fb12339d0e/python/sglang/srt/models/spark2_5.py#L198-L206)

```python
if self.headwise_attn_output_gate:
    g, _ = self.g_proj(hidden_states)
    g = torch.sigmoid(g.float()).to(attn_output.dtype)
    gate_output = attn_output.view(...) * g.unsqueeze(-1)
    attn_output = gate_output.view(*attn_output.shape)
```

The activation function is **not** hardcoded in the reference. The official `modeling_spark.py` selects it from the config field `gate_attn_act_mode`, supporting `sigmoid` and `silu`, and raises on anything else:

[`modeling_spark.py`](https://huggingface.co/XHToken/Spark-X2.5-4B/blob/main/modeling_spark.py) (lines 199-204)

```python
if self.gate_attn_act_mode == "sigmoid":
    gate = torch.sigmoid(gate_score.float())
elif self.gate_attn_act_mode == "silu":
    gate = F.silu(gate_score.float())
else:
    raise ValueError(f"Unsupported gate_attn_act_mode: {self.gate_attn_act_mode}")
```

The two functions differ substantially: `sigmoid(x) ∈ (0, 1)` is bounded and strictly positive, while `silu(x) = x · sigmoid(x)` is unbounded and goes negative for `x < 0`. Because the gate is multiplied into every head of every attention layer's output, the error accumulates layer over layer and the model output diverges from the reference. SGLang also has no `else: raise`, so an unsupported mode would fail silently rather than erroring at load time.

### Current impact

Both released checkpoints — [Spark-X2.5-4B](https://huggingface.co/XHToken/Spark-X2.5-4B/blob/main/config.json) and [Spark-X2.5-1.7B](https://huggingface.co/XHToken/Spark-X2.5-1.7B/blob/main/config.json) — set `gate_attn_act_mode: "sigmoid"`, so the hardcoded value happens to match and there is **no numerical divergence on the released models today**. I verified this by reading both configs.

This is a latent/forward-compatibility mismatch rather than an active regression: any future Spark checkpoint declaring `gate_attn_act_mode: "silu"` (the reference explicitly supports it) would be served with the wrong gate activation and produce garbage output, with no error to point at the cause. It is the same class of "hardcode-the-known-config-value, drop the config read" omission as #37609 (GELU approximation), but in the attention path.

### Reproduction

The activation divergence is visible on CPU without loading model weights:

```python
import torch
import torch.nn.functional as F

x = torch.linspace(-6, 6, 20001, dtype=torch.float32)

# SGLang hardcoded path
sglang_gate = torch.sigmoid(x)

# reference path for a "silu" checkpoint
ref_gate = F.silu(x)

print("SGLang (hardcoded sigmoid) range:", sglang_gate.min().item(), sglang_gate.max().item())
print("reference (silu) range:        ", ref_gate.min().item(), ref_gate.max().item())
print("max abs diff:                  ", (sglang_gate - ref_gate).abs().max().item())
```

Observed output:

```text
SGLang (hardcoded sigmoid) range: 0.002472617670893669 0.997527361869812
reference (silu) range:         -0.3171774744987488 6.0
max abs diff:                   5.4314165115356445
```

For the released `sigmoid` checkpoints the two paths coincide (`max abs diff = 0.0`), consistent with "no current divergence, latent mismatch".

### Expected behavior

Read `gate_attn_act_mode` from the config in `Spark2_5Attention.__init__` and branch on it in `forward`, mirroring the reference's three branches including the `else: raise`. Something like:

```python
# __init__
self.gate_attn_act_mode = getattr(config, "gate_attn_act_mode", "sigmoid")

# forward
if self.gate_attn_act_mode == "sigmoid":
    g = torch.sigmoid(g.float())
elif self.gate_attn_act_mode == "silu":
    g = F.silu(g.float())
else:
    raise ValueError(f"Unsupported gate_attn_act_mode: {self.gate_attn_act_mode}")
g = g.to(attn_output.dtype)
```

### Environment

Reported against `main` at `f8cbf000f4a5bfd86d3fb7c1e2d6c8fb12339d0e`. No GPU run needed — the mismatch is in the activation function selection, verifiable from config + source alone.

**Comment by ActiveSky** [comment]
Hi @Bevisy, thanks for the detailed report and the CPU-only analysis.

We checked the current upstream main, and Spark2_5Attention still appears to hardcode torch.sigmoid without reading gate_attn_act_mode. We would like to try a focused config-aware fix for the supported sigmoid and silu modes, with validation for unsupported values and CPU unit coverage.

We noticed that PR #37637 also touches python/sglang/srt/models/spark2_5.py for the exact-GELU issue. To avoid stepping on that work, we can keep the changes separate and rebase or coordinate as appropriate.

We are community contributors, so please let us know if you or another contributor are already handling this, or if the maintainers prefer a different scope. If there is no overlap, we will prepare a PR from the latest upstream main. Thank you.

**Comment by Bevisy** [comment]
No worries, that is my fix pr 😄

**Comment by ActiveSky** [comment]
Thanks for clarifying, @Bevisy! We understand that #37637 is your fix PR for the exact GELU behavior.

Our work is focused on a separate issue, #37639: making `Spark2_5Attention` honor the configured `gate_attn_act_mode` (preserving the current `sigmoid` behavior while supporting the configured `silu` path), together with focused CPU regression tests. Although both changes touch `spark2_5.py`, we are keeping the scopes separate and can rebase or coordinate as needed.

Please let us know if you notice any overlap or if the maintainers prefer a particular merge order. Thanks again!
